### PR TITLE
ci: add Dockerfile

### PR DESCRIPTION
**Pull request by hqwuzhaoyi**
I plan to add a Dockerfile for possible local deployment needs.

The temporary process is as follows

1. Change the `.env.template` file

2. docker build
```
docker build -t shiro . --load       
```
3. docker run
```
docker run --name shiro -d -p 2233:3000 shiro
```

I studied the next packaging and temporarily cannot put the NEXT env in the environment variables of docker run. I can only package it locally first.

CleanShot 2023-08-04 at 17 35 58@2x

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+FROM base AS runner
+WORKDIR /app
+
+COPY --from=builder /app/public ./public
+COPY --from=builder /app/.next/standalone ./
+COPY --from=builder /app/.next/static ./.next/static
+COPY --from=builder /app/.next/server ./.next/server
+
+EXPOSE 3000
```
Comment (Innei) [inline Dockerfile:32]:
```suggestion
EXPOSE 2323
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+
+WORKDIR /app
+COPY --from=deps /app/ .
+RUN npm install -g pnpm
+RUN pnpm build
+
+FROM base AS runner
+WORKDIR /app
+
```
Comment (Innei) [inline Dockerfile:26]:
```suggestion

ENV NODE_ENV production
ARG BASE_URL
ENV BASE_URL=${BASE_URL}
ENV NEXT_PUBLIC_API_URL=${BASE_URL}/api/v2
ENV NEXT_PUBLIC_GATEWAY_URL=${BASE_URL}

# and other docker env inject
```

**Review by Innei — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 2 above)
Comment (Innei) [reply Dockerfile:26]: Refer: https://github.com/mx-space/kami/blob/master/Dockerfile

**Comment by hqwuzhaoyi** [comment]
@Innei I just tried it, but in the next `appDir` mode, the environment variables need to be generated on page `build`. If they are not placed in env before building, packaging will not work properly.

CleanShot 2023-08-07 at 10 51 02

I think we can create the .env file first, or put ARG before Build.

**Comment by hqwuzhaoyi** [comment]
> @Innei I just tried it, but in the next `appDir` mode, the environment variables need to be generated on page `build`. If they are not placed in env before building, packaging will not work properly.
> 
> CleanShot 2023-08-07 at 10 51 02
> 
> I think we can create the .env file first, or put ARG before Build.

I will commit it immediately.

**Comment by hqwuzhaoyi** [comment]
@Innei 
I added a docker-compose.yml.

Now just need to change the args inside.

Then execute 

```
docker compose up -d
```

or use docker command

```
docker build \
 --build-arg BASE_URL=BASE_URL \
 --build-arg NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY=NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY \
 --build-arg CLERK_SECRET_KEY=CLERK_SECRET_KEY \
 -t shiro . --load
```

```
docker run --name shiro -d -p 2323:2323 shiro
```

**Review by Innei — APPROVED** [review]
Thank you!

**Pull request merged**